pagefaultgames/pokerogue
# [Enhancement] Improvements to starter selection and filtering user experience

## What are the changes?
Change the generation filter to an "hybrid" filter that allows both to pick any combination of gens like currently and to switch quickly from one gen to the other like before filters were introduced.
A good amount of refactoring, adding documentation and improvements on the "backend" filter code was involved as well.

## Why am I doing these changes?
Following PR #2916  there was a lot of negative feedback about how navigating through all Pokemon with the new UI took way too many inputs compared to previously.
By adding this hybrid behavior we hopefully have the best of both worlds: fast navigation between gens while still being able to filter specific gens in any way we want to.
A good amount of refactoring was pretty much needed to do the changes as the previous implementation for the filters left a lot of margin for error when creating filters. The changes should allow us to add more filters in the future way more easily.

## What did change?

### New hybrid type of filtering
* When all options in this filter are unselected, the option that is currently in focus will be used to filter instead of nothing. While in this state, moving from an option to the next with up/down will allow to quickly go through all options without any extra input.
* When selecting any option or more the behavior is the same as before
* The generation, type and shiny filters now all use this filtering
* For the generation filter, by default all is unselected and the cursor is on generation 1
* The generation filter has been moved to the left of the filter bar so that when going through the generations in this way the Pokemon are not hidden by the filters, effectively allowing to go through all generations quickly, as before the filters were introduced
* For the type and shiny filters, this behavior is not enabled by default

### UI updates
* Make the Filter Bar fill all available horizontal space
* Make the Starter Team fill up all available vertical space
* Update arrow navigation between filter bar, team and possible starters to work with these changes
* Automatically move cursor to the first starter when closing the filters
* Fix the coloring of modified filters for legacy theme
* Set the color of enabled filter options to blue instead of green to avoid green/red combination for color blindness
* Set the size of the multi state filters from PR #3252 to the widest of options in the filters, instead of having the container resize itself when toggling the filter
* Rename the Filter categories based on future plans for filters and add missing localization strings

### Other QoL stuff
* Reset all filters to default when creating a new game. Before the previous filters were kept for the whole session
* For challenges, all generations are selected by default instead of just generation 1. Along with PR #3312 all valid Pokemon are now properly shown when creating a new challenge

### "Backend" filter improvements that will help for future updates
* Helper functions to select or unselect all elements in a filter category
* Update the toggle filters introduced in PR #3252 to allow for any number of options instead of just 3
* Filter options can now display both a sprite and text, while before it was only one or the other

## Screenshots/Videos

(all before videos and screenshots are taken on current beta)

### UI update

Before (default theme):
<img width="798" alt="_ui-default-before" src="https://github.com/user-attachments/assets/d5ee5da4-a520-4adb-9d10-07574e1ddd13">

After (default theme):
<img width="798" alt="_ui-default-after" src="https://github.com/user-attachments/assets/a0ec20a7-9d12-4671-bdca-1600c138672e">

Before (legacy theme):
<img width="1166" alt="_ui_legacy_before" src="https://github.com/user-attachments/assets/5111cff7-4ae7-4b09-91ab-865cc49e2a2f">

After (legacy theme):
<img width="800" alt="_ui-legacy-after" src="https://github.com/user-attachments/assets/f51fcadd-b2c1-42de-8710-7245f2bcb681">


### New filtering for Generation, Type and  Shiny

Before: trying to go through one gen at a time

https://github.com/user-attachments/assets/8050f0f0-cd60-4917-b232-9f9bb50de493

After: going through one gen at a time

https://github.com/user-attachments/assets/5f913d61-ebe0-4db7-895d-81218db6f211

After: showcasing the hybrid behavior of the new filter type

https://github.com/user-attachments/assets/b51c89d0-10c3-4c78-9203-05e5049fa50a


### Filter reset when creating new game

Before:

https://github.com/user-attachments/assets/0be4907b-ea9f-4649-b2b7-3cf99dd9257a

After

https://github.com/user-attachments/assets/a61259a2-1205-4875-9b88-9711f81c2e90

### Multi toggle filters no longer change size

Before:

https://github.com/user-attachments/assets/9d5faca9-093f-408a-9c0a-d45eef23be13

After:

https://github.com/user-attachments/assets/31f5610c-60df-4cdf-8145-cc6346d0cdf6


## How to test the changes?

- Create a new game in any mode and play around with the filters
- Try moving around between the UI elements in various states of filtering and team composition
- Exit and go back to creating a new game to make sure filters are properly reset

I haven't added unit tests for the DropDown and FilterBar classes, but it's definitely something that would be beneficial and that I'd like to add hopefully in the near future. But for now I'd rather have all the QoL improvements live regardless of those tests being in, if possible. Even moreso because some of those changes are blocking further updates to the filters mentioned

## Future Work

Putting here for reference, as this PR is already way too big for its own good

- Add more filters for Pokerus, hidden ability, and other types of unlocks [(see what's planned here)](https://discord.com/channels/1125469663833370665/1267339517371875361/1268014543511224360)
- Add shortcut to open the filter menu from anywhere
- When in filter menu, add shortcuts to select all and unselect all elements
- Update the look of hybrid filter to make the current state clearer
- Probably more that I forget and will add when I remember :') 

## Checklist
- [X] The PR is self-contained and cannot be split into smaller PRs? (not really due to heavy refactoring)
- [ ] Have I considered writing automated tests for the issue? (considered? yes. done? no)
